packages/alfa-rules/src/sia-r16/rule.ts

```diff
@@ -32,53 +35,175 @@ export default Rule.Atomic.of<Page, Element>({
       },
 
       expectations(target) {
+        const diagnostic = hasRequiredValues(device, target);
+
         return {
           1: expectation(
-            hasRequiredValues(device)(target),
-            () => Outcomes.HasAllStates,
-            () => Outcomes.HasNotAllStates
+            diagnostic.isOk(),
+            () => Outcomes.HasAllStates(diagnostic.get()),
+            () => Outcomes.HasNotAllStates(diagnostic.getErr())
           ),
         };
       },
     };
   },
 });
 
-function hasRequiredValues(device: Device): Predicate<Element> {
-  return (element) => {
-    const node = Node.from(element, device);
-
-    for (const role of node.role) {
-      // The `separator` role is poorly architected in the sense that its
-      // inheritance and attribute requirements depend on aspects of the element
-      // carrying the role. If the element is not focusable, the `separator`
-      // role has no required attributes.
-      if (role.is("separator") && !isFocusable(device)(element)) {
-        return true;
-      }
+function hasRequiredValues(
+  device: Device,
+  element: Element
+): Result<RoleAndRequiredAttributes> {
+  let result = true;
+
+  const node = Node.from(element, device);
+
+  let roleName: string = "";
+  let required: Array<aria.Attribute.Name> = [];
+  let missing: Array<aria.Attribute.Name> = [];
+  let found: Array<aria.Attribute.Name> = [];
+
+  for (const role of node.role) {
+    roleName = role.name;
+    // The `separator` role is poorly architected in the sense that its
+    // inheritance and attribute requirements depend on aspects of the element
+    // carrying the role. If the element is not focusable, the `separator`
+    // role has no required attributes.
+    if (role.is("separator") && !isFocusable(device)(element)) {
+      return Ok.of(
+        RoleAndRequiredAttributes.of("", roleName, required, missing, found)
+      );
+    }
 
-      for (const attribute of role.attributes) {
-        if (
-          role.isAttributeRequired(attribute) &&
-          node.attribute(attribute).every(property("value", isEmpty))
-        ) {
-          return false;
-        }
+    required = Array.copy(role.requiredAttributes());
+
+    for (const attribute of required) {
+      if (node.attribute(attribute).every(property("value", isEmpty))) {
+        missing.push(attribute);
+        result = false;
+      } else {
+        found.push(attribute);
       }
     }
+  }
```

Comment:
Since we end up using `role.requiredAttributes` only here (rather than twice: here and in the expectation), I think it's better to roll back to the original version so that we're not leaking the stuff about `separator` being a mess in two places.
```suggestion
    for (const attribute of role.attributes) {
      if (role.isAttributeRequired(attribute)) {
        if (node.attribute(attribute).every(property("value", isEmpty))) {
          missing.push(attribute);
          result = false;
        } else {
          found.push(attribute);
        }
      }
    }
  }
```